### PR TITLE
[PW-5426] donations webapi endpoint

### PR DESCRIPTION
**Pull request by peterojo**
**Description**

Add /donations webapi endpoint
Create donations request components
Trigger donation request from webapi
Receive and process API response.

***Note:*** More testing and response processing would have to be done when implementing the adyen giving web component in the checkout success page

**Tested scenarios**

Mock request through postman, triggers chain of request builder -> API request -> response validator and handler.
Throws exceptions on error.

### CODE REVIEW

File: etc/acl.xml [hunk 1]
```diff
@@ -9,8 +9,9 @@
                         <resource id="Adyen_Payment::guestPaymentMethods" title="Get payment methods for guest shoppers" sortOrder="20" />
                         <resource id="Adyen_Payment::paymentMethods" title="Get payment methods for logged-in shoppers" sortOrder="30" />
                         <resource id="Adyen_Payment::paymentDetails" title="Submit details for a payment" sortOrder="40" />
-                        <resource id="Adyen_Payment::stateData" title="Send state data in a seperate request" sortOrder="50" />
+                        <resource id="Adyen_Payment::stateData" title="Send state data in a separate request" sortOrder="50" />
                         <resource id="Adyen_Payment::paymentStatus" title="Get order payment status" sortOrder="50" />
+                        <resource id="Adyen_Payment::donations" title="Adyen Giving" translate="title" sortOrder="60"/>
```
Comment (acampos1916) [inline etc/acl.xml:14]:
```suggestion
                        <resource id="Adyen_Payment::donations" title="Submit Adyen Giving request" translate="title" sortOrder="60"/>
```

**Review by acampos1916 — APPROVED** [review]

File: Gateway/Http/Client/TransactionDonate.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+     */
+    public function placeRequest(TransferInterface $transferObject)
+    {
+        $request = $transferObject->getBody();
+        $service = new Checkout($this->client);
+
+        try {
+            $response = $service->donations($request);
+        } catch (AdyenException $e) {
+            $response['error'] = $e->getMessage();
```
Comment (michaelpaul) [inline Gateway/Http/Client/TransactionDonate.php:61]:
```suggestion
            $response = array( 'error' => $e->getMessage());
```

File: Gateway/Request/DonationDataBuilder.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+     * @param array $buildSubject
+     * @return array
+     * @throws NoSuchEntityException
+     */
+    public function build(array $buildSubject)
+    {
+        $request['body'] = $this->adyenRequestsHelper
+            ->buildDonationData($buildSubject['payment'], $this->storeManager->getStore()->getId());
+
+        return $request;
```
Comment (michaelpaul) [inline Gateway/Request/DonationDataBuilder.php:60]:
```suggestion
        return array(
            'body' => $this->adyenRequestsHelper
                ->buildDonationData($buildSubject['payment'], $this->storeManager->getStore()->getId())
        );
```

File: Gateway/Validator/DonateResponseValidator.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        $response = SubjectReader::readResponse($validationSubject);
+
+        if (empty($response['resultCode'])) {
+            $errorMsg = __('An error occurred with the donation.');
+
+            if (!empty($response['error'])) {
+                $this->adyenLogger->error($response['error']);
+            }
+
+            throw new LocalizedException(__($errorMsg));
```
Comment (michaelpaul) [inline Gateway/Validator/DonateResponseValidator.php:62]:
```suggestion
            if (!empty($response['error'])) {
                $this->adyenLogger->error($response['error']);
            }
            throw new LocalizedException(__('An error occurred with the donation.'));
```

File: Model/Api/AdyenDonations.php [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+     * @throws NotFoundException
+     */
+    public function donate($payload)
+    {
+        $payload = json_decode($payload, true);
+
+        // Validate JSON that has just been parsed if it was in a valid format
+        if (json_last_error() !== JSON_ERROR_NONE) {
+            throw new LocalizedException(__('Donations call failed because the request was not a valid JSON'));
+        }
```
Comment (michaelpaul) [inline Model/Api/AdyenDonations.php:58]: Can we use https://devdocs.magento.com/guides/v2.4/extension-dev-guide/framework/serializer.html here?

**Review by michaelpaul — CHANGES_REQUESTED** [review]
Some small tweaks and questions.

File: Gateway/Validator/DonateResponseValidator.php (hunk 4 above)
Comment (peterojo) [reply Gateway/Validator/DonateResponseValidator.php:62]: Good catch, I already updated this in #1218

File: etc/webapi.xml [hunk 6]
```diff
@@ -109,4 +109,18 @@
...
     </route>
+
+    <route url="/V1/adyen/donations" method="POST">
+        <service class="Adyen\Payment\Api\AdyenDonationsInterface" method="donate"/>
+        <resources>
+            <resource ref="Adyen_Payment::donations"/>
+        </resources>
+    </route>
+
+    <route url="/V1/internal/adyen/donations" method="POST">
```
[the remark this answers is not in the document]
Comment (peterojo) [reply etc/webapi.xml:120]: I'll investigate this option and do it in a different ticket if it's possible

File: Gateway/Http/Client/TransactionDonate.php (hunk 2 above)
Comment (peterojo) [reply Gateway/Http/Client/TransactionDonate.php:61]: Changed in 6a1f3e7

File: Gateway/Request/DonationDataBuilder.php (hunk 3 above)
Comment (peterojo) [reply Gateway/Request/DonationDataBuilder.php:60]: changed in 6a1f3e7

File: Model/Api/AdyenDonations.php (hunk 5 above)
Comment (peterojo) [reply Model/Api/AdyenDonations.php:58]: Added in cee73a0

**Comment by peterojo** [comment]
@michaelpaul I added all the updates in #1218 , that branch is based on this one and will be merged together to the feature/adyen-giving branch

File: etc/webapi.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (peterojo) [reply etc/webapi.xml:120]: worth investigating but will be a separate issue if applicable

[the remark this answers is not in the document]
Comment (Morerice) [reply etc/webapi.xml:120]: This was attempted, but if a regular controller (like `Json.php`) was used we would have to parse the post parameters one by one either in the constructor or in the `execute` function, whereas trough `webapi` we can include these parameters directly in the `handleInternalRequest` arguments

**Review by Morerice — APPROVED** [review]

**Review by michaelpaul — APPROVED** [review]

**Pull request merged**